Add support for Mediatek mt8195 platform

yaochunhung opened the pull request:
Hi, 
This PR is for support mt8195 platform on SOF.
MT8195 include Cadence HiFi-4 single core, currently we support 4 DAI interfaces, SOF_DL2/DL3/UL4/UL5(Downlink/Uplink path) on MT8195 Audio Hardware(which we call AFE, the Abbreviation of Audio Front End). Some ASoC codes(the committer is Trevor) in the PR is accepted by ASoC reviewer but not in the sof-linux branch, please ignore them then help to review on SOF part.
Thanks!

paulstelian97 commented on sound/soc/sof/pcm.c:
```diff
@@ -913,7 +913,8 @@ void snd_sof_new_platform_drv(struct snd_sof_dev *sdev)
 	pd->pcm_construct = sof_pcm_new;
-	pd->ignore_machine = drv_name;
+	if (!plat_data->desc->no_ignore_machine)
+		pd->ignore_machine = drv_name;
```
This change doesn't look backward compatible (default for new fields is supposed to be a 0-like value, in this case false).

yaochunhung replied:
default is zero and set as true if want to reuse original platform driver. we set it as true in the sof_of_mt8195_desc. Is there anything for my misunderstanding? Please advice.Thanks!

plbossart replied:
I would go with @dbaluta's suggestion. Call the flag the same in your descriptor as in the ASoC driver definition, this will avoid logical inversions that are unfortunately oh so common in engineering circles.

ranj063 commented on include/sound/sof/dai-mediatek.h:
```diff
@@ -0,0 +1,26 @@
+#ifndef __INCLUDE_SOUND_SOF_DAI_MEDIATEK_H__
+#define __INCLUDE_SOUND_SOF_DAI_MEDIATEK_H__
+
+//#include <linux/types.h>
```
please remove this.

yaochunhung replied:
ok. I will remove it.

ranj063 commented on include/sound/sof/dai-mediatek.h:
```diff
@@ -0,0 +1,26 @@
+
+struct sof_ipc_dai_afe_params {
+	struct sof_ipc_hdr hdr;
+
```
unncessary new line

yaochunhung replied:
ok. I will remove it.

ranj063 commented on include/sound/sof/dai-mediatek.h:
```diff
@@ -0,0 +1,26 @@
+	u32 rate;
+	u32 format;
+	u32 stream_id;
+
```
same here

yaochunhung replied:
ok. I will remove it.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195-clk.c:
```diff
@@ -0,0 +1,198 @@
+
+int adsp_enable_clock(struct device *dev)
+{
+	int ret = 0;
```
unnecessary init

yaochunhung replied:
ok, I will modify it.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195-clk.c:
```diff
@@ -0,0 +1,198 @@
+
+	ret = clk_prepare_enable(clk_handle[CLK_TOP_MAINPLL_D7_D2]);
+	if (ret) {
+		dev_info(dev, "%s clk_prepare_enable(mainpll_d7_d2) fail %d\n",
```
shouldnt this be dev_err? and all of the below?

yaochunhung replied:
ok, I will modify them. thanks!

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195-clk.c:
```diff
@@ -0,0 +1,198 @@
+	if (ret) {
+		dev_info(dev, "%s clk_prepare_enable(audio_h_sel) fail %d\n",
+			 __func__, ret);
+		return ret;
```
no need for this. you can just print the error and let the return ret at the end take of this.

yaochunhung replied:
ok, I will remove it.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195-clk.c:
```diff
@@ -0,0 +1,198 @@
+
+int adsp_clock_on(struct device *dev)
+{
+	int ret = 0;
```
no need to init

yaochunhung replied:
ok, I will remove it.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195-clk.c:
```diff
@@ -0,0 +1,198 @@
+	if (ret) {
+		dev_err(dev, "%s [ADSP] adsp_default_clk_init fail ret : %d\n",
+			__func__, ret);
+		return ret;
```
please remove this return

yaochunhung replied:
ok, I will remove it.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195-clk.c:
```diff
@@ -0,0 +1,198 @@
+
+int adsp_clock_off(struct device *dev)
+{
+	int ret = 0;
```
unnecessary init

yaochunhung replied:
ok. I will modify it.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195-clk.c:
```diff
@@ -0,0 +1,198 @@
+	if (ret) {
+		dev_err(dev, "%s [ADSP] adsp_default_clk_init fail ret : %d\n",
+			__func__, ret);
+		return ret;
```
remore this

yaochunhung replied:
ok. I will remove it.

plbossart commented:
> Some ASoC codes(the committer is Trevor) in the PR is accepted by ASoC reviewer but not in the sof-linux branch, please ignore them then help to review on SOF part.

What patches are you referring to? We merge upstream code weekly (that includes Mark Brown and Takashi Iwai for-next branches), so everything needed should be there?

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,376 @@
+
+static int platform_parse_resource(struct platform_device *pdev, void *data)
+{
+	int ret = 0;
```
remove init

ranj063 replied:
and move after the struct declarations please

yaochunhung replied:
ok, I will modify it.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,376 @@
+	struct device_node *mem_region;
+	struct device *dev = &pdev->dev;
+	struct adsp_chip_info *adsp = data;
+	int i = 0;
```
remove init

yaochunhung replied:
ok, I will modify it.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,376 @@
+	res = &resource;
+	mem_region = of_parse_phandle(dev->of_node, "memory-region", 0);
+	if (!mem_region) {
+		dev_info(dev, "no 'memory-region' phandle\n");
```
dev_err?

yaochunhung replied:
ok, I will modify it

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,376 @@
+
+	ret = of_address_to_resource(mem_region, 0, res);
+	if (ret) {
+		dev_info(dev, "of_address_to_resource failed\n");
```
err?

ranj063 replied:
please fix this everywhere else too please. there are a lof ot dev_info's in place of dev_err

yaochunhung replied:
ok. I will modify them

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,376 @@
+/*  Init the basic DSP DRAM address */
+static int adsp_memory_remap_init(struct device *dev)
+{
+	int err = 0;
```
err is not used at all.

yaochunhung replied:
ok. I will remove it.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,376 @@
+	int ret = 0;
+	u32 shared_size;
+	struct device *dev = &pdev->dev;
+	struct adsp_chip_info *adsp = data;
```
reverse xmas tree style for delcarations please

yaochunhung replied:
ok. I will modify it.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,376 @@
+static int mt8195_dsp_probe(struct snd_sof_dev *sdev)
+{
+	struct platform_device *pdev =
+		container_of(sdev->dev, struct platform_device, dev);
```
will fit in 100chars on one line?

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,376 @@
+	struct platform_device *pdev =
+		container_of(sdev->dev, struct platform_device, dev);
+	struct adsp_priv *priv;
+	int ret = 0, mailbox_type;
```
no init

yaochunhung replied:
ok, I will modify it.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,376 @@
+	ret = adsp_sram_power_on(sdev->dev, true);
+	if (ret) {
+		dev_err(sdev->dev, "adsp_sram_power_on fail!\n");
+		ret = -EINVAL;
```
why modify the return value here?

yaochunhung replied:
ok. I will keep the return value.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,376 @@
+	ret =  adsp_memory_remap_init(&pdev->dev);
+	if (ret) {
+		dev_err(sdev->dev, "adsp_memory_remap_init fail!\n");
+		ret = -EINVAL;
```
same here

yaochunhung replied:
ok. I will keep the return value.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,376 @@
+	ret =  adsp_shared_base_ioremap(pdev, adsp_info);
+	if (ret) {
+		dev_err(sdev->dev, "adsp_shared_base_ioremap fail!\n");
+		ret = -EINVAL;
```
same here

yaochunhung replied:
ok. I will keep the return value.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,376 @@
+	sdev->mailbox_bar = mailbox_type;
+
+err_pm_disable:
+	pm_runtime_disable(&pdev->dev);
```
I could be wrong but don;t you need to do a runtime_put() and then disable?

plbossart replied:
@ranj063 is right, but you don't need a get_sync() in the first place.

The premise of pm_runtime is that you probe with all resources allocated. If you enable pm_runtime pm, and do nothing, after some time the resources will be suspended, and if you need them the ALSA framework will do a get_sync()  and resume the resources as needed. You do not need to increment the refcount in your probe function.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,376 @@
+static int mt8195_dsp_remove(struct snd_sof_dev *sdev)
+{
+	struct platform_device *pdev =
+		container_of(sdev->dev, struct platform_device, dev);
```
one line

plbossart commented on include/sound/sof.h:
```diff
@@ -68,6 +68,9 @@ struct sof_dev_desc {
 	bool use_acpi_target_states;
 
+	/*set to 1 to not use ignore_machine and still use the original platform driver*/
+	bool no_ignore_machine;
```
unclear why you need this for?

yaochunhung replied:
Because we need to reuse original platform driver in our design, we don't want to overwrite it in the function soc_check_tplg_fes to still use original platform. Then we can route sof link to original(normal sink), 
e.g.  sof_conn_stream : { "ETDM2_OUT_BE", "AFE_SOF_DL2", SOF_DMA_DL2, SNDRV_PCM_STREAM_PLAYBACK}, to route sof_link : SOF_DMA_DL2 to normal link ETDM2_OUT_BE, then we can use normal link setting for audio hardware.

dbaluta replied:
I think going with `ignore_machine` name here works better.

ranj063 replied:
@yaochunhung I totally missed this. ignore_machine has nothing to do with ignoring the machine driver selected in SOF. It is more to do with how the core deals with the FE DAI links in the legacy machine driver. In your case, setting this should have no effect at all because there's no legacy driver to deal with right? @plbossart is my understanding correct?

yaochunhung replied:
@ranj063 We still use legacy driver because some pcm nodes are still controlled by non-sof driver, not move to DSP. And We also want to keep to legacy BE DAI driver because the power and clock of our audio hardware are controlled by AP side. e.g ETDM2_OUT_BE. If no_ignore_machine is set as true, the legacy driver can be keep and reuse. Thanks!

yaochunhung replied:
I plot the block diagram to explain our driver. (1). Normal ALSA pcm platform driver, (2). SOF ALSA pcm platform driver. Sorry about that I don't know how to draw it in the github but hope it more clear. Thanks!
````
  ||====================================================||       
  ||(1).Normal ALSA pcm platform ////////////////////////////////////////////////////////||                                                
  || -------------- ////////////////////////////////////////////////////////---------------////||                                      
  || | DLx................| ///////////////////////////////////////////////////////| I2S playback|////||                                      
  || |(normal FE) | ---------------------------------------------------> | (normal BE)..|////||                                      
  ||  --------------- ///////////////////////////////////////////////////////---------------////||  
  || ///////////////////////////////////////////////////////////////////////////////////////////||                                      
  || /////////////////////////////////////////--------------------///////-------------------///||                                               
  ||///////////////////////////////////////// | SOF_DMA_DL2 |/////// | ETDM2_OUT_BE |//||         							             	                     
  || /////////////////////////////////////////|....(widget)........... |-----> | (normal BE)..........|//||                                      
  ||///////////////////////////////////////// -------------------////////--------------------//||                                      
  || //////////////////////////////////////////////////// ^/////////////////////////////////////||                                      
  ||===============================|=====================||                                     
                        Connect SOF DL2 to SOF BE CPU DAI Widget’s source widget 
  ||===============================|=====================||                                     
  ||(2).SOF ALSA pcm platform//////////////////////...|...///////////////////////////////////||                                     
  ||////////////////////////////////////////////////////....|..////////////////////////////////////||                                     
  ||/------------///////----------------/////////////--------------/////----------------//////||                                     
  ||/| SOF_DL2|----->| PCMxP............|--->......--->| BUFn............|--->| AFE_SOF_DL2|//////||                                       
  ||/| (SOF FE).|///////|(Host widget)|/////////////|(buf widget)|/////| (SOF BE)..........|//////||                                     
  ||/ -----------///////---------------//////////////--------------//////----------------//////||                                     
  ||////////////////////////////////////////////////////////////////////////////////////////////||                                     
  ||====================================================|| 
````

dbaluta replied:
@yaochunhung can you draw this into something more user friendly? :)

yaochunhung replied:
@dbaluta Sorry that I don't know how to attach file or diagram to github website. May I send the diagram to you, @plbossart, and @ranj063  dieclty by e-mail? Thanks!

plbossart replied:
use e.g. https://asciiflow.com/#/ and copy-paste the diagram with backquotes to protect it.

yaochunhung replied:
@plbossart  Thanks but I still have problem to copy-paste the diagram with backquotes. I share the diagram in the link. Thanks. https://asciiflow.com/#/share/eJzlWN2K2kAUfpVDrhQWL4Sl1Lv4xy5oV9BlbwIymtkaGichjlVZFvoQZaGv0cvSp8mT7Ex%2B3JlkEpOo3dYOQ0jGk%2FN952fOmfikEbTEWousbftKs9EOe1pLezK0raG1Pl5fXxnajt01P%2FA7ireUPRia%2F%2F33RU7DIOwKAPpgrMPYWRMTOsgzocJ4U%2FfuZp1wymZV8UvKSymVnxxviewwBu58Ca6N6CNbO0LlGVieXmWVcBzCKB1cBSmuqjvYiszZ2FjmZ0zBc9YUg4voQoS9bY5hxKrJDM2%2F7G1Vqq5F0e736pU8UHS%2B%2FGLXGKzNwXJo%2FR1hOJE%2FzpC1kKntrHj62rQc6HsOodAj5tnxaiHgDesAG%2BThugJvsrBW8VbYWLYNMwzINLEJFimNpxaQ8JZovrAIBtOzvmLvCohDAT0%2B4jmNpAtn7FFZWTU1U%2FaQcENKZT60rRG5Z3zXn3aH%2BrQ7aO7Re5PusDm9u59M2z1QcotypavfAuKZosJWO%2FzlpxgcORR%2BXEjUUZQZdLGLGfTNgyRUgEG%2BB2thtsUFLLhIdc3PPHKENbDDto%2Fn2DbL0dmuSm%2F7s1kmHQeYgeX5piOrMun0uV2MVY6ozEotcKFnzCI2A8tkwiufUCKAOkCRx8sxW2X5QjujeybMC8FDsHH8bz9WsHLW3hzHhVvAK81V5HOKmEOCTVTHhnpQSBKI7x6qM0XdT5aptN1Migc4qAqjzjDzIyHP12nUQCfLlboUibPu6pSoCrVkQh4SiWCOUyPrS1tVLvoKi8N9zfa0uClYrLcj0W3spn3fJ43tm5De703jdzP4xb4We4KCgRwWPsP2X7txVlRcbDT2P9baa3Yg8zLPCwcYJR6SjGo8Q%2Fs9qf1Hla0u%2BkRa46zCcljPYnD8%2FEfzFTLdrYpIuXEQ5GJmjpkZuV3Fl0qAi5wKU%2FOcccS4bD%2BmvOmfOi%2BTr%2B8BgqPEeJSQKMi1sGHlPJBB1bVcbPO%2FFXL0BX0sanvqxgdR4xN%2FDgD54buMVVkz%2B5zNljvO0nUIJlT0IexXhW6Zlsw%2FxL97yMp%2BX%2BQP%2Fz%2FY8Ib2rD2%2FAhvlfrM%3D)

dbaluta replied:
![image](https://user-images.githubusercontent.com/62316/132476344-923dfe3a-5305-43e5-9fc8-c63d9ab2c58f.png)

plbossart replied:
Sorry, I have absolutely zero understanding of the suggested solution @yaochunhung 

The most puzzling part is the 'ETDM2_OUT_BE' which isn't going to be connected to the 'SOF_DL2 FE'. So when you start the 'SOF DL2 FE', how would e.g. the HW_PARAMS be propagated to the ETDM2_OUT_BE which controls power and clock?

yaochunhung replied:
@plbossart  In the machine driver sound/soc/mediatek/mt8195/mt8195-mt6359-rt1019-rt5682.c. we define g_sof_conn_streams about the connection among ETDM2_OUT_BE,SOF_DMA_DL2 and AFE_SOF_DL2. 
static const struct sof_conn_stream g_sof_conn_streams[] = {
	{ "ETDM2_OUT_BE", "AFE_SOF_DL2", SOF_DMA_DL2, SNDRV_PCM_STREAM_PLAYBACK},
	{ "ETDM1_OUT_BE", "AFE_SOF_DL3", SOF_DMA_DL3, SNDRV_PCM_STREAM_PLAYBACK},
	{ "UL_SRC1_BE", "AFE_SOF_UL4", SOF_DMA_UL4, SNDRV_PCM_STREAM_CAPTURE},
	{ "ETDM2_IN_BE", "AFE_SOF_UL5", SOF_DMA_UL5, SNDRV_PCM_STREAM_CAPTURE},
};
Then we implement .late_probe = mt8195_mt6359_rt1019_rt5682_card_late_probe. 
In this function, we search pcm_runtime by sof_link and normal_link then get sof's pcm runtime : sof_rtd and normal pcm runtime : normal_rtd. After that, we search the sof's runtime cpu_dai then check the widget in its path then set dapm route's source and sink then add routes.

for example, ETDM2_OUT_BE is normal_link, AFE_SOF_DL2 is sof_link,SOF_DMA_DL2 is sof_dma, and stream_dir is SNDRV_PCM_STREAM_PLAYBACK. 
We will use snd_soc_dapm_widget_for_each_source_path to get the playback_widget is BUFn.x in the diagram above.
then set sink of dpam route as conn->sof_dma(SOF_DMA_DL2 in this case). After that, BUFn.x will be connected to SOF_DMA_DL2.
For SOF_DMA_DL2 to ETDM2_OUT_BE , we define the dapm route path in machine driver.
	/* SOF Downlink */
	{"I070", NULL, SOF_DMA_DL2},
	{"I071", NULL, SOF_DMA_DL2},
And also define dapm route path in mt8195-dai-etdm.c
	{"O048", "I070 Switch", "I070"},
	{"O049", "I071 Switch", "I071"},
	{"ETDM2 Playback", NULL, "O048"},
	{"ETDM2 Playback", NULL, "O049"},
I hope this explanation is more clear. Thanks.

plbossart replied:
> I hope this explanation is more clear. Thanks.

Humm, no :-(
I am missing the big picture, you just hammered us with details that are probably correct but don't really provide a sense of what the problem statement is, if you considered alternatives and what led you to pick this solution you're asking us to review.

yaochunhung replied:
@plbossart  
We set no_ignore_machine as true to prevent ignoring normal FE in soc_check_tplg_fes.
 
/* ignore this FE */
if (dai_link->dynamic) {
dai_link->ignore = true;
continue;
})
 
The reason is that we still need to use normal FE for the paths which don't gothrough DSP. The benifit is that we can offload the process of certain audio data(e.g. speaker/headphone) to dsp and we can reuse normal alsa driver for those audio data which don't need dsp offloading.(e.g. Display port). Thanks.

ranj063 replied:
@yaochunhung It makes a bit more sense now. but a naive question. Is there a way for you define the Normal FE-> Normal BE in topology?

yaochunhung replied:
@ranj063  For normal FE/BE, I think it need to communicate with DSP if defining them in topology(please correct me if my misunderstanding).  It takes lower effort to reuse normal FE/BE without topology. Thanks.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -28,23 +28,17 @@
-
-	return adsp->va_cfgreg;
-}
-EXPORT_SYMBOL(get_adsp_reg_base);
```
unrelated change?

yaochunhung replied:
yes, I will  modify it.

plbossart commented on sound/soc/sof/mediatek/Kconfig:
```diff
@@ -0,0 +1,22 @@
+
+config SND_SOC_SOF_MTK_TOPLEVEL
+	bool "SOF support for MTK audio DSPs"
+	depends on ARM64|| COMPILE_TEST
```
ARM64 || COMPILE_TEST

plbossart replied:
doesn't this also depends on the SOF_OF config at teh higher level?

yaochunhung replied:
Yes, I will add SND_SOC_SOF_OF also, thanks!

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -361,12 +375,16 @@ struct snd_sof_dsp_ops sof_mt8195_ops = {
 	.block_write	= sof_block_write,
 
-	/* Module IO */
+	/* Register IO */
```
should be in the previous patch?

yaochunhung replied:
yes, I will modify it.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -234,6 +234,19 @@ static int adsp_shared_base_ioremap(struct platform_device *pdev, void *data)
+static int mt8195_run(struct snd_sof_dev *sdev)
+{
+	u32 adsp_bootup_addr;
+	int ret = 0;
```
ret not used

yaochunhung replied:
ok. I will remove it.

ranj063 commented on sound/soc/sof/mediatek/adsp_pcm.c:
```diff
@@ -0,0 +1,246 @@
+int adsp_genpool_create(struct adsp_mem_pool *mem_pool,
+			struct adsp_mem *mem_info)
+{
+	int ret = 0;
```
remove int

plbossart replied:
remove initialization

yaochunhung replied:
ok, I will remove it. thanks

plbossart commented on sound/soc/sof/mediatek/adsp_helper.h:
```diff
@@ -0,0 +1,57 @@
+#define __ADSP_HELPER_H__
+
+#include <linux/of_irq.h>
+#include <linux/platform_device.h>
```
move those to where you include this file? I don't see any dependency on platform_device below.

yaochunhung replied:
yes, they are unnecessary and I will remove it.

yaochunhung replied:
ok. I will remove them. Thanks!

ranj063 commented on sound/soc/sof/mediatek/adsp_pcm.c:
```diff
@@ -0,0 +1,246 @@
+int adsp_genpool_alloc(struct adsp_mem_pool *mem_pool, unsigned char **vaddr,
+		       dma_addr_t *paddr, unsigned int size)
+{
+	*vaddr = NULL;
```
is this really needed?

yaochunhung replied:
I think it is unnecessary because gen_pool_dma_alloc will return NULL if alloc fail and I will remove it.

ranj063 commented on sound/soc/sof/mediatek/adsp_pcm.c:
```diff
@@ -0,0 +1,246 @@
+			     unsigned char **vaddr, dma_addr_t *paddr,
+			     unsigned int size, int align)
+{
+	*vaddr = NULL;
```
same here

yaochunhung replied:
ok. I will remove it

plbossart commented on sound/soc/sof/mediatek/mt8195/mt8195-clk.c:
```diff
@@ -0,0 +1,198 @@
+	}
+
+	clk_handle[CLK_TOP_AUDIO_LOCAL_BUS_SEL] = devm_clk_get(dev,
+							       "audio_local_bus");
```
one line if it fits on 100 chars?

yaochunhung replied:
yes, I will modify it

plbossart commented on sound/soc/sof/mediatek/mt8195/mt8195-clk.c:
```diff
@@ -0,0 +1,198 @@
+	}
+
+	clk_handle[CLK_SCP_ADSP_AUDIODSP] = devm_clk_get(dev,
+							 "scp_adsp_audiodsp");
```
one line?

yaochunhung replied:
ok. I will modify it. thanks

ranj063 commented on sound/soc/sof/mediatek/adsp_pcm.c:
```diff
@@ -0,0 +1,246 @@
+	/* allocate VA with gen pool */
+	if (*vaddr) {
+		gen_pool_free(mem_pool->dram_pool, (unsigned long)vaddr, size);
+		size = 0;
```
is is needed?

plbossart replied:
this has no effect!

yaochunhung replied:
ok. I will remove it.

plbossart commented on sound/soc/sof/mediatek/mt8195/mt8195-clk.c:
```diff
@@ -0,0 +1,198 @@
+
+int adsp_enable_clock(struct device *dev)
+{
+	int ret = 0;
```
useless init

yaochunhung replied:
ok, I will modify it.

ranj063 commented on sound/soc/sof/mediatek/adsp_pcm.c:
```diff
@@ -0,0 +1,246 @@
+		ret = adsp_genpool_alloc(mem_pool,
+					 &pg_table->area,
+					 &pg_table->addr,
+					 pg_table->bytes);
```
use 100chars per line please

yaochunhung replied:
ok.thanks

plbossart commented on sound/soc/sof/mediatek/mt8195/mt8195-clk.c:
```diff
@@ -0,0 +1,198 @@
+		adsp_disable_clock(dev);
+	}
+
+TAIL:
```
unusual label, what's wrong with 'err:'

yaochunhung replied:
It will show error log if return err. Should it return err directly?

ranj063 commented on sound/soc/sof/mediatek/adsp_pcm.c:
```diff
@@ -0,0 +1,246 @@
+	struct adsp_priv *priv = sdev->pdata->hw_pdata;
+	struct adsp_mem_pool *mem_pool = &priv->mem_pool;
+	size_t dma_bytes;
+	int ret = 0;
```
i think this init is also not required. please check

yaochunhung replied:
yes, it is unnecessary. I will remove it.

plbossart commented on sound/soc/sof/mediatek/mt8195/mt8195-clk.c:
```diff
@@ -0,0 +1,198 @@
+			goto TAIL;
+		}
+	} else {
+		adsp_disable_clock(dev);
```
no error check?

yaochunhung replied:
There is no return value from adsp_disable_clock function.

ranj063 commented on sound/soc/sof/mediatek/adsp_pcm.c:
```diff
@@ -0,0 +1,246 @@
+
+	return 0;
+out:
+	if (dmab) {
```
dmab cannot be NULL here right? no need to check?

yaochunhung replied:
yes, it shouldn"t be NULL and I will remove the check.

plbossart commented on sound/soc/sof/mediatek/mt8195/mt8195-clk.c:
```diff
@@ -0,0 +1,198 @@
+
+int adsp_clock_on(struct device *dev)
+{
+	int ret = 0;
```
useless init.

yaochunhung replied:
ok, I will remove it.

plbossart commented on sound/soc/sof/mediatek/mt8195/mt8195-clk.c:
```diff
@@ -0,0 +1,198 @@
+
+int adsp_clock_off(struct device *dev)
+{
+	int ret = 0;
```
useless init.

yaochunhung replied:
ok. I will modify it.

plbossart commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,376 @@
+#include <linux/of_irq.h>
+#include <linux/pm_domain.h>
+#include <linux/delay.h>
+#include <linux/of_reserved_mem.h>
```
alphabetical order?

yaochunhung replied:
ok, I will modify it.

plbossart commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,376 @@
+
+static void *get_adsp_chip_data(void)
+{
+	return (void *)adsp_info;
```
is the cast required?

yaochunhung replied:
I think yes because adsp_info is declaration as struct adsp_chip_info *. please correct if my misundertanding.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -358,6 +358,52 @@ static int mt8195_dsp_remove(struct snd_sof_dev *sdev)
+{
+	struct platform_device *pdev =
+		container_of(sdev->dev, struct platform_device, dev);
+	int ret = 0;
```
no init

yaochunhung replied:
ok, I will modify it.

plbossart commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,376 @@
+
+static int platform_parse_resource(struct platform_device *pdev, void *data)
+{
+	int ret = 0;
```
useless uinit

yaochunhung replied:
ok, I will modify it.

yaochunhung commented:
> What patches are you referring to? We merge upstream code weekly (that includes Mark Brown and Takashi Iwai for-next branches), so everything needed should be there?
Below are mt8195 ASoC commits in the linux kernel.
https://git.kernel.org/pub/scm/linux/kernel/git/broonie/sound.git/log/?qt=author&q=Trevor+Wu
Because I add the commit https://github.com/thesofproject/linux/pull/3122/commits/ea950f5ef4a6daf2002458077fd25e85b7fa34e3 is related with machine driver. So I also add them into the PR. Thanks!

plbossart commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,376 @@
+	struct device_node *mem_region;
+	struct device *dev = &pdev->dev;
+	struct adsp_chip_info *adsp = data;
+	int i = 0;
```
try to use reverse x-mas tree structure, with long declarations first.

yaochunhung replied:
ok, I will modify it.

plbossart commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,376 @@
+		 resource_size(res));
+
+	ret = of_reserved_mem_device_init(dev);
+
```
newline doesn't seem aligned with the style in the rest of this function.

yaochunhung replied:
ok I will align them. thanks!

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -358,6 +358,52 @@ static int mt8195_dsp_remove(struct snd_sof_dev *sdev)
+	adsp_sram_power_on(&pdev->dev, false);
+
+	/* power down DSP */
+	pm_runtime_put_sync(&pdev->dev);
```
this part is a bit confusing. Can you please explain why you need the put_sync()here?

plbossart replied:
it's wrong. using pm_runtime INSIDE of suspend/resume functions makes no sense. You cannot suspend while you suspend.

yaochunhung replied:
There are still some problems that I cannot do sof dsp boot flow without get_sync so I use put_sync in the suspend function. But it is abnormal method so I will remove it first. Then check how to add PM later by @dbaluta 's suggestion. Thanks.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -358,6 +358,52 @@ static int mt8195_dsp_remove(struct snd_sof_dev *sdev)
+	int ret = 0;
+
+	/* power on DSP */
+	pm_runtime_get_sync(&pdev->dev);
```
same here, why do you need the get_sync() here?

plbossart replied:
completely wrong, sorry.

yaochunhung replied:
I will remove it first then check how to add PM later. Thanks.

plbossart commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,376 @@
+	writel(offset, vaddr_emi_map);
+	dev_info(dev, "After vaddr_emi_map 0x%x\n", readl(vaddr_emi_map));
+
+	return err;
```
return 0? You are not using err at all.

yaochunhung replied:
ok. I will modify it.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -358,6 +358,52 @@ static int mt8195_dsp_remove(struct snd_sof_dev *sdev)
+	ret = adsp_clock_on(&pdev->dev);
+	if (ret) {
+		dev_err(sdev->dev, "adsp_clock_on fail!\n");
+		return -EINVAL;
```
return ret please

yaochunhung replied:
ok. thanks!

plbossart commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,376 @@
+	}
+	dev_info(dev, "[ADSP] shared-dram vbase=%p, phy addr :%llx,  size=0x%x\n",
+		 adsp->shared_dram, adsp->pa_shared_dram, shared_size);
+tail:
```
err:

yaochunhung replied:
ok. I will modify it.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -358,6 +358,52 @@ static int mt8195_dsp_remove(struct snd_sof_dev *sdev)
+	ret = adsp_sram_power_on(sdev->dev, true);
+	if (ret) {
+		dev_err(sdev->dev, "[ADSP] adsp_sram_power_on fail!\n");
+		return -EINVAL;
```
remove this. return ret at the end will handle it

yaochunhung replied:
ok. I will remove it. thanks!

plbossart commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,376 @@
+	/* remap shared-dram base to be non-cachable */
+	shared_size = TOTAL_SIZE_SHARED_DRAM_FROM_TAIL;
+	//adsp->pa_shared_dram = adsp->pa_dram + DSP_DRAM_SIZE ;
+	adsp->pa_shared_dram = adsp->pa_dram + adsp->dramsize - shared_size;
```
remove comment above

yaochunhung replied:
ok. I will remove it.

plbossart commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,376 @@
+	struct platform_device *pdev =
+		container_of(sdev->dev, struct platform_device, dev);
+	struct adsp_priv *priv;
+	int ret = 0, mailbox_type;
```
useless init

yaochunhung replied:
ok, I will modify it.

plbossart commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,376 @@
+	}
+
+	pm_runtime_enable(&pdev->dev);
+	pm_runtime_get_sync(&pdev->dev);
```
why would you do this? this will prevent the device for suspending?

plbossart replied:
this is still problematic!

plbossart commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,376 @@
+		return -EINVAL;
+	}
+
+	pm_runtime_enable(&pdev->dev);
```
enable pm_runtime LAST, when all the probe succeeded

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195-ipc.c:
```diff
@@ -0,0 +1,240 @@
+		return ret;
+	}
+
+	// init value;
```
use /* */ fr comments please

yaochunhung replied:
ok. I will fix it. Thanks!

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -54,6 +55,41 @@ void __iomem *get_mbox_reg_base(u32 id)
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: failed to init adsp ipi device:%d\n",
+			ret);
+		return ret;
```
remove this

yaochunhung replied:
ok. I will remove it. Thanks!

plbossart commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,376 @@
+
+	adsp_clock_off(&pdev->dev);
+	adsp_sram_power_on(&pdev->dev, false);
+	pm_runtime_put_sync(&pdev->dev);
```
no, that's not good at all. essentially you've disable pm_runtime with this, the device will NEVER suspend.

you need still need to disable pm_runtime in the remove function, mirroring the enable in the probe

plbossart commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -28,23 +28,17 @@
 static struct adsp_chip_info *adsp_info;
 
+static struct snd_soc_acpi_mach sof_mt8195_mach = {
+	.id = "819501",
```
Is this an ACPI HID? This is clearly not following the standard if yes, there should be a vendor identification.

yaochunhung replied:
No, we use acpi mach for mt8195, this id is mt'8195' with version '01'

plbossart replied:
````
@id: ACPI ID (usually the codec's) used to find a matching machine driver.
````
-> where is this "819501" in the codec driver?

yaochunhung replied:
No, it is not codec driver. id is useless. we use the drv_name"mt8195_mt6359_rt1019_rt5682" in the structure for sof_machine_register.

plbossart commented on sound/soc/sof/mediatek/mt8195/mt8195-loader.c:
```diff
@@ -0,0 +1,69 @@
+				DSP_RESET_SW, DSP_RESET_SW);
+
+	/* toggle  Dreset & BReset */
+	/* pull high DReset & BReset */
```
Dreset or DReset, be consistent with your spelling.

yaochunhung replied:
DReset, thanks!

plbossart commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -234,6 +234,19 @@ static int adsp_shared_base_ioremap(struct platform_device *pdev, void *data)
+		 adsp_bootup_addr);
+	hifixdsp_boot_sequence((void *)sdev, adsp_bootup_addr);
+
+	return ret;
```
return 0

yaochunhung replied:
ok. I will modify it.

plbossart commented on sound/soc/sof/mediatek/adsp_helper.h:
```diff
@@ -9,6 +9,8 @@
 
 #include <linux/of_irq.h>
 #include <linux/platform_device.h>
+#include "adsp_pcm.h"
```
you're including this twice, remove here.

yaochunhung replied:
Do you mean adsp_pcm.h? I only see once.

plbossart commented on sound/soc/sof/mediatek/adsp_pcm.c:
```diff
@@ -0,0 +1,246 @@
+		return -EINVAL;
+
+	*vaddr = (unsigned char *)gen_pool_dma_alloc(mem_pool->dram_pool,
+			size, paddr);
```
remove initialization at the top of function

yaochunhung replied:
ok. I will remove it.

plbossart commented on sound/soc/sof/mediatek/adsp_pcm.c:
```diff
@@ -0,0 +1,246 @@
+		}
+
+		kfree(dmab);
+	}
```
else?

yaochunhung replied:
we don't genpool if managed_buffer_alloc is false so do nothing for else

plbossart commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -358,6 +358,52 @@ static int mt8195_dsp_remove(struct snd_sof_dev *sdev)
+	/*turn off adsp clock*/
+	adsp_clock_off(&pdev->dev);
+
+	/*power down adsp sram*/
```
use spaces after '/* ' and before ' */'

yaochunhung replied:
ok. thanks!

plbossart commented on sound/soc/sof/mediatek/adsp_ipc.c:
```diff
@@ -0,0 +1,64 @@
+	struct adsp_mbox_ch_info *ch_info = priv->ipi[idx].chan->con_priv;
+
+	ch_info->ipi_op_val = op;
+	mbox_send_message(priv->ipi[idx].chan, NULL);
```
no error checks?

yaochunhung replied:
ok, I will check return value. thanks!

plbossart commented on sound/soc/sof/mediatek/mt8195/mt8195-ipc.c:
```diff
@@ -0,0 +1,240 @@
+	sdev = ch_info->priv->sdev;
+	pdev = container_of(sdev->dev, struct platform_device, dev);
+
+	name[4] = '0' + ch_info->id;
```
use kasprintf or something, it's easier to review

yaochunhung replied:
ok. I will use kasprintf. Thanks!

plbossart commented on sound/soc/sof/sof-of-dev.c:
```diff
@@ -50,6 +51,16 @@ static struct sof_dev_desc sof_of_imx8mp_desc = {
 #endif
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_MT8195)
+static struct sof_dev_desc sof_of_mt8195_desc = {
+	.no_ignore_machine = true,
```
what does this mean?

yaochunhung replied:
This flag use to bypass ignore_machine so that we can still use no-SOF platform driver.

plbossart commented on sound/soc/mediatek/mt8195/mt8195-mt6359-rt1019-rt5682.c:
```diff
@@ -25,6 +25,21 @@
 #define RT5682_DEV0_NAME	"rt5682.2-001a"
 
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_MT8195)
+#include <sound/sof.h>
```
why would you include SOF-related stuff in a machine driver? This is a layering violation.

yaochunhung replied:
OK. I will use #include <sound/soc-acpi.h> for snd_soc_acpi_mach. Thank!

plbossart commented on sound/soc/mediatek/mt8195/mt8195-mt6359-rt1019-rt5682.c:
```diff
@@ -698,6 +737,169 @@ SND_SOC_DAILINK_DEFS(UL_SRC2_BE,
+{
+	int i, j, ret = 0;
+	struct snd_soc_card *card = rtd->card;
+	struct snd_soc_dai_link *sof_dai_link = NULL;
```
this is likely wrong, you would need to reset the dai_link at every iteration for the for(i loop

yaochunhung replied:
ok. I will modify it. Thanks!

plbossart commented on sound/soc/mediatek/mt8195/mt8195-mt6359-rt1019-rt5682.c:
```diff
@@ -698,6 +737,169 @@ SND_SOC_DAILINK_DEFS(UL_SRC2_BE,
+
+static int mt8195_mt6359_rt1019_rt5682_card_late_probe(struct snd_soc_card *card)
+{
+	int i = 0;
```
initialization not needed, and move two lines below.

yaochunhung replied:
ok. I will modify it.

plbossart commented on sound/soc/mediatek/mt8195/mt8195-mt6359-rt1019-rt5682.c:
```diff
@@ -698,6 +737,169 @@ SND_SOC_DAILINK_DEFS(UL_SRC2_BE,
+{
+	int i = 0;
+	struct snd_soc_pcm_runtime *runtime = NULL;
+	struct snd_soc_component *sof_comp = NULL;
```
initializations not needed.

yaochunhung replied:
ok. I will modify them.

plbossart commented on sound/soc/mediatek/mt8195/mt8195-mt6359-rt1019-rt5682.c:
```diff
@@ -698,6 +737,169 @@ SND_SOC_DAILINK_DEFS(UL_SRC2_BE,
+		const struct sof_conn_stream *conn = &g_sof_conn_streams[i];
+		struct snd_soc_pcm_runtime *sof_rtd = NULL;
+		struct snd_soc_pcm_runtime *normal_rtd = NULL;
+		struct snd_soc_pcm_runtime *rtd = NULL;
```
revisit all initializations please.

yaochunhung replied:
ok. I will modify it.

plbossart commented on sound/soc/mediatek/mt8195/mt8195-mt6359-rt1019-rt5682.c:
```diff
@@ -698,6 +737,169 @@ SND_SOC_DAILINK_DEFS(UL_SRC2_BE,
+	for_each_card_prelinks(card, i, dai_link) {
+		if (dai_link->no_pcm && !dai_link->stream_name && dai_link->name)
+			dai_link->stream_name = dai_link->name;
+	}
```
what is this doing? This looks like ASoC code that's duplicated in the machine driver?

yaochunhung replied:
https://elixir.bootlin.com/linux/latest/source/sound/soc/soc-core.c#L1851 here  skip assign stream_name in soc_check_tplg_fes function when no_ignore_machineis true,  so we assign them in the mt8195_mt6359_rt1019_rt5682_card_probe

plbossart replied:
Not understanding the explanation, sorry.

plbossart commented on sound/soc/mediatek/mt8195/mt8195-mt6359-rt1019-rt5682.c:
```diff
@@ -998,12 +1231,25 @@ static int mt8195_mt6359_rt1019_rt5682_dev_probe(struct platform_device *pdev)
+#else
 	platform_node = of_parse_phandle(pdev->dev.of_node,
 					 "mediatek,platform", 0);
+#endif
```
this doesn't look good to me, if there is a single build where you want deal with two platforms, you need to auto-detect the platform then make a selection at run time. Compile-time selections are really problematic for distributions. @cujomalainey please comment here.

cujomalainey replied:
Sorry missed this comment. I am just trying to make sure I understand this code correctly. Its a compile time switch to select which driver services the platform correct? @yaochunhung does MTK plan to make SOF the default for this platform? Or do you plan to use the other driver still? @plbossart this seems like a reasonable switch for during the transition period, much like we had on GLK for a while no?

yaochunhung replied:
@cujomalainey Yes, we plan to make SOF the default for this platform. thanks!

plbossart commented:
@yaochunhung You will need to work with @dbaluta to update the ARM64 Kconfig so that we can compile your code in GitHub CI. See https://github.com/thesofproject/kconfig/blob/master/kconfig-sof-arm64.sh

dbaluta commented:
@yaochunhung I will provide you more feedback on Monday. Care to share parts of your dts file? Also, do you reuse your non-SOF machine drivers right?

yaochunhung reviewed (commented):
Update patch for ASoC: mediatek: mt8195: add sof to machine driver

plbossart commented on include/sound/sof/dai-mediatek.h:
```diff
@@ -0,0 +1,23 @@
+
+#include <sound/sof/header.h>
+
+struct sof_ipc_dai_afe_params {
```
mtk_afe_params?

we'll need to start using prefixes now that we've got more contributors - which is a good thing!

yaochunhung replied:
yes. I will modify it

plbossart commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,370 @@
+
+	adsp_clock_off(&pdev->dev);
+	adsp_sram_power_on(&pdev->dev, false);
+	pm_runtime_put_sync(&pdev->dev);
```
no that's wrong. You should disable pm_runtime here.

yaochunhung replied:
OK. Currently I still check the pm_runtime flow because I need pm_runtime_get_sync to power on adsp power and clock. I need them to read/write dsp register. Therefore, I need to get_sync to power on adsp power/clock when load firmware, run firmware, and even for IPC.  I am still finding a better method for it. Thanks!

yaochunhung replied:
@plbossart I need to power on adsp power and clock in the probe function because I need to access dsp registers for some initialization(e.g power on DSP SRAM). Therefore I use get/put sync to ensure not increasing refcount in the probe function. I also do get /put sync paired in these functions : load firmware (get/put sync) , dsp_pre_fw_run(get sync), and machine_register(put sync, I don't put sync in the post_fw_run because I also need adsp power in init_trace, not sure if it is better to get sync in init_trace then put sync in release trace).  With these modification, pm_runtime refcount will be zero after sof_probe function.  But I encounter one problem is that I still need adsp power/clock when load topology but adsp power is power down. Not sure if there is anything I missed or misunderstanding. Could you please give some advice? Thanks a lot!

plbossart replied:
No, that's wrong to auto-resume your own driver. the pm_runtime_get_sync() is used e.g. by ASoC to make sure the DSP is resumed when you start playing audio. It's a utility for *OTHERS* to request the DSP be back in full power mode.

Again, the flow is that on probe you power everything, then mark the DSP as idle and let it suspend if no one else is using the DSP. That determination is done by the pm_runtime framework, not your driver.

yaochunhung replied:
Thanks for advise. Is it ok to use get/put sync for sof probe flow? I need to power on adsp power and clock in the probe function because I need to access dsp registers for some initialization(e.g power on DSP SRAM). Therefore I use get/put sync to ensure not increasing refcount in the probe function. I also do get /put sync paired in these functions : load firmware (get/put sync) , dsp_pre_fw_run(get sync), and machine_register(put sync, I don't put sync in the post_fw_run because I also need adsp power in init_trace, not sure if it is better to get sync in init_trace then put sync in release trace). With these modification, pm_runtime refcount will be zero after sof_probe function. But I encounter one problem is that I still need adsp power/clock when load topology but adsp power is power down. Not sure if there is anything I missed or misunderstanding. Could you please give some advice? Thanks a lotI need to power on adsp power and clock in the probe function because I need to access dsp registers for some initialization(e.g power on DSP SRAM). Therefore I use get/put sync to ensure not increasing refcount in the probe function. I also do get /put sync paired in these functions : load firmware (get/put sync) , dsp_pre_fw_run(get sync), and machine_register(put sync, I don't put sync in the post_fw_run because I also need adsp power in init_trace, not sure if it is better to get sync in init_trace then put sync in release trace). With these modification, pm_runtime refcount will be zero after sof_probe function. But I encounter one problem is that I still need adsp power/clock when load topology but adsp power is power down. Not sure if there is anything I missed or misunderstanding. Could you please give some advice? Thanks a lot!

plbossart commented on sound/soc/sof/mediatek/adsp_pcm.c:
```diff
@@ -0,0 +1,237 @@
+//
+
+#include "adsp_pcm.h"
+#include "adsp_helper.h"
```
"you're including this twice, remove here."

this is the inclusion, so you don't need to have #include "adsp_pcm.h" in "adsp_helper.h"

yaochunhung replied:
ok. I will remove it. thanks

plbossart reviewed (changes requested):
Thanks @yaochunhung for a lot of improvements. Major blocking points remain on the pm_runtime usage and the 'ignore_machine' requirements.

dbaluta commented on include/sound/sof/dai-mediatek.h:
```diff
@@ -0,0 +1,23 @@
+/* SPDX-License-Identifier: (GPL-2.0-only OR BSD-3-Clause) */
+/*
+ *Copyright(c) 2021 Mediatek Corporation. All rights reserved.
+ *
```
Add a space between `*` and `Copyright`.

yaochunhung replied:
ok. I will add it. thanks!

dbaluta commented on sound/soc/sof/mediatek/mt8195/mt8195-clk.c:
```diff
@@ -0,0 +1,192 @@
+	if (ret) {
+		dev_err(dev, "%s clk_prepare_enable(scp_adsp_audiodsp) fail %d\n",
+			__func__, ret);
+		return ret;
```
If clk_prepare_enable fails here we should call clk_disable_unprepare for the rest of the clocks.

yaochunhung replied:
ok. I will modify it. thanks!

dbaluta commented on sound/soc/sof/Kconfig:
```diff
@@ -226,5 +226,6 @@ config SND_SOC_SOF_PROBE_WORK_QUEUE
 source "sound/soc/sof/imx/Kconfig"
```
@yaochunhung would love to see some more details in the commit message. E.g How is the hardware configuration for your platform, what kind of architecture / cores is on application processor, what kind of DSP do you have. 
Also some information about the way these two entities communicate, how are the resources managed (clocks, power domains, pinmuxing, etc).
This is useful for review.

yaochunhung replied:
ok. I will add more information. Thanks!

dbaluta commented on sound/soc/sof/mediatek/adsp_helper.h:
```diff
@@ -0,0 +1,58 @@
+	u32 (*ap2adsp_addr)(u32 addr);
+	u32 (*adsp2ap_addr)(u32 addr);
+
+	void *private_data;
```
Also can you provide a short memory map from AP view and DSP view. I suspect there is no 1 to 1 mapping. We have the same scenarios on one of our boards and we've implemented similar functions with ap2adsp_addr but on the firmware side not on the kernel. Not sure which is better. I dont suggest to change this if it fits your needs, but it must be explained in some comments or in the commit message.

yaochunhung replied:
There is a memory remapping for DRAM between AP view and DSP view. In DSP view, the start address of EMI is 0x50000000 and in AP view is 0x40000000. In current design, I set dram of SOF between AP and DSP are the same. Both are 0x60000000. the memory remapping is zero, Could you please tell me what's the keyword for your implementation? I will check it. Thanks!

dbaluta commented on sound/soc/sof/mediatek/mt8195/mt8195-clk.c:
```diff
@@ -0,0 +1,192 @@
+
+	va_dspsysreg = devm_ioremap(dev, ADSP_SRAM_POOL_CON, 0x4);
+
+	if (!va_dspsysreg) {
```
No need from new line here. Between function call and return value check.

yaochunhung replied:
ok. I will remove it. thanks.

dbaluta commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -29,23 +29,17 @@
-	adsp = get_adsp_chip_data();
-	if (!adsp)
-		return NULL;
-
```
You shouldn't delete code that was previously added in a patch of the same Pull Request. Just do not add it at all from the beginning.

yaochunhung replied:
ok. I will modify it.

dbaluta commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -345,6 +339,26 @@ static int mt8195_dsp_remove(struct snd_sof_dev *sdev)
+
+	/* get machine node and save it to mach->pdata */
+	mach->pdata = of_get_child_by_name(sdev->dev->of_node, "sound");
+	if (!mach->pdata)
```
Definitely this needs more explanations. How to you plan to describe your machine. Maybe create finer grained patches with proper patches explanations. This should also go to review to DT maintainer.

dbaluta commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -339,6 +351,16 @@ static int mt8195_dsp_remove(struct snd_sof_dev *sdev)
+#endif
+	return type;
+}
+
```
This function here is not related with this commit. Maybe add it in the initial commit with hardware support. Also, need to explain the use of iffdef.

yaochunhung replied:
ok. I will move it in the initial commit. Not ADSP_USE_DRAM is only used in early development stage for DSP SRAM only case, But DSP SRAM is only 256KB, I think it won't be used without DRAM. I think the definition can be removed. Thanks.

dbaluta commented:
@yaochunhung I've gone your patch series and added some comments. There is still work to be done. I would suggest you take it slower with simple patches first and then go for more delicate ones. 

You can run git log --oneline sound/soc/sof/imx/ and see how we added our initial support for i.MX8 boards. It is not complete, we are now sending our clocks patches, then following by PM support, then the rest of the stuff. 

If you want and consider it necessary we could also setup a meeting to discuss more.

cujomalainey commented:
Should the commit `dt-ndings: mediatek: mt8195: add dsp document` be sent to ALSA out of this PR?

ranj063 commented on sound/soc/sof/topology.c:
```diff
@@ -803,6 +804,19 @@ static const struct sof_topology_token led_tokens[] = {
 
+/* AFE */
+static const struct sof_topology_token afe_tokens[] = {
+	{SOF_TKN_MEDIATEK_AFE_RATE,
```
how does this patch compile? you dont introduce the tokens until a later patch?

yaochunhung replied:
I define it in the patch https://github.com/thesofproject/linux/pull/3122/commits/87e7086623800eaa6fb7c6a7a722f87403bc1e9e#diff-9c75bd60b09aeda0d73593bb94af4f39ad560d6a80aef2c0a957c049cc240b9eR144. Thanks!

ranj063 replied:
hmm maybe it is the ordering of the patches as seen on github. I was just wondering if the patch that defines the tokens is before or after this one

yaochunhung replied:
Maybe I should reorder the patches to define these tokens before using them. Thanks!

ranj063 commented on sound/soc/sof/topology.c:
```diff
@@ -2892,6 +2906,55 @@ static int sof_link_esai_load(struct snd_soc_component *scomp, int index,
+			       ARRAY_SIZE(afe_tokens), private->array,
+			       le32_to_cpu(private->size));
+	if (ret != 0) {
+		dev_err(scomp->dev, "error: parse afe tokens failed %d\n",
```
you can drop the "error" prefix from this error and all the others too.

yaochunhung replied:
ok. I will modify them.thanks!

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195-clk.c:
```diff
@@ -0,0 +1,192 @@
+	if (ret) {
+		dev_err(dev, "%s clk_prepare_enable(audio_h_sel) fail %d\n",
+			__func__, ret);
+	}
```
unncessary braces

yaochunhung replied:
ok, I will modify it.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195-clk.c:
```diff
@@ -0,0 +1,192 @@
+			dev_err(dev, "failed to adsp_enable_clock: %d\n", ret);
+			goto TAIL;
+		}
+	} else {
```
you can remove the else and return ret inside the if {} at the end. the goto is unncessary really. You can just have return ret instead.

yaochunhung replied:
else is used for disable clock so I think it should be kept.

ranj063 replied:
what I meant was having a return inside the if(enable) so you wont need the else. but its only minor

yaochunhung replied:
May I know if it looks like that? I will modify it. Thanks!
                if (ret)
                        dev_err(dev, "failed to adsp_enable_clock: %d\n", ret);

                return ret;
          }

         adsp_disable_clock(dev);

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195-clk.c:
```diff
@@ -0,0 +1,192 @@
+	if (ret) {
+		dev_err(dev, "%s [ADSP] adsp_default_clk_init fail ret : %d\n",
+			__func__, ret);
+	}
```
remove braces

ranj063 replied:
I think you can simply call "return adsp_default_clk_init(dev, 1);" in this function. The failures are already printed in adsp_default_clk_init()

yaochunhung replied:
ok. I will modify it.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195-clk.c:
```diff
@@ -0,0 +1,192 @@
+{
+	int ret = 0;
+
+	dev_info(dev, "%s: %s\n", __func__, enable ? "on" : "off");
```
should this be dev_dbg?

yaochunhung replied:
ok.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195-clk.c:
```diff
@@ -0,0 +1,192 @@
+	if (ret) {
+		dev_err(dev, "%s [ADSP] adsp_default_clk_init fail ret : %d\n",
+			__func__, ret);
+	}
```
same here. just return adsp_default_clk_init(dev, 0);

yaochunhung replied:
ok. I will modify it.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,370 @@
+	if (!adsp)
+		return NULL;
+
+	if (id >= DSP_MBOX_NUM)
```
i think this check should be first

yaochunhung replied:
ok. I will modify it.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,370 @@
+	struct adsp_chip_info *adsp = data;
+	int i, ret;
+
+	res = &resource;
```
whats the idea here? The variable resource is not really used anywhere?

yaochunhung replied:
It is only used to parse dts information then store to adsp_chip_info. Then we can get adsp_chip_info  to use them after all.

ranj063 replied:
Ahh I see what you're doing.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,370 @@
+	dev_info(dev, "%s: 0x%llx\n", __func__, adsp->pa_dram);
+
+	if (!adsp)
+		return -EBUSY;
```
why -EBUSY?

yaochunhung replied:
Is ENXIO better for it?

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,370 @@
+	int offset;
+
+	adsp = get_adsp_chip_data();
+	dev_info(dev, "%s: 0x%llx\n", __func__, adsp->pa_dram);
```
this is misplaced. What if adsp is NULL?

yaochunhung replied:
Yes. adsp check should be before this log but I think the log can be removed.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,370 @@
+		 offset);
+	WARN_ON(offset < 0);
+	writel(offset, vaddr_emi_map);
+	dev_info(dev, "After vaddr_emi_map 0x%x\n", readl(vaddr_emi_map));
```
you have a lot of dev_info's. do they need to info or can they be dbg level?

yaochunhung replied:
These information is used to ensure the dram address and offset between host and dsp. So I suggest to keep them.

ranj063 replied:
right im not asking to remove. just wondering if they all need to be info level or can they be dev_dbg()

yaochunhung replied:
I remove some and keep this. thanks

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,370 @@
+		adsp->shared_dram = devm_ioremap(dev, adsp->pa_shared_dram,
+						 shared_size);
+		if (!adsp->shared_dram) {
+			dev_info(dev, "ioremap failed at line %d\n", __LINE__);
```
dev_err please

yaochunhung replied:
ok. I will modify it. thanks.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,370 @@
+			goto err;
+		}
+	}
+	dev_info(dev, "[ADSP] shared-dram vbase=%p, phy addr :%llx,  size=0x%x\n",
```
dbg?

yaochunhung replied:
This log is used to show shared buffer between AP and dsp, So I suggest to keep it. Thanks.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,370 @@
+	ret = platform_parse_clock(&pdev->dev);
+	if (ret) {
+		dev_err(sdev->dev, "platform_parse_clock failed\n");
+		return -EINVAL;
```
return ret

yaochunhung replied:
ok

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,370 @@
+
+	ret = platform_parse_resource(pdev, adsp_info);
+	if (ret) {
+		dev_err(sdev->dev, "platform_parse_resource failed\n");
```
i'm sure you don;t need this error message. platform_parse_resource() should already print the appropraite error right?

yaochunhung replied:
yes. I will remove it. thanks.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,370 @@
+
+	ret = platform_parse_clock(&pdev->dev);
+	if (ret) {
+		dev_err(sdev->dev, "platform_parse_clock failed\n");
```
same here, do you need this message?

yaochunhung replied:
ok. I will remove it. thanks.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,370 @@
+
+	ret =  adsp_memory_remap_init(&pdev->dev);
+	if (ret) {
+		dev_err(sdev->dev, "adsp_memory_remap_init fail!\n");
```
have you checked the error paths? if there's an error here, do you take care or powering off the SRAM and clock?

yaochunhung replied:
yes, I will add clock and sram power off. thanks.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,370 @@
+
+	sdev->mmio_bar = mailbox_type;
+	sdev->mailbox_bar = mailbox_type;
+
```
you would end up disabling PM even in case of a successful probe? is this intentional? Looks like the Pm handling needs a bit of careful consideration

yaochunhung replied:
ok. I will check the correct PM flow.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -345,6 +339,26 @@ static int mt8195_dsp_remove(struct snd_sof_dev *sdev)
+	/* get machine node and save it to mach->pdata */
+	mach->pdata = of_get_child_by_name(sdev->dev->of_node, "sound");
+	if (!mach->pdata)
+		dev_warn(sdev->dev, "warning: get child machine node \"sound\" failed\n");
```
remove warning prefix

yaochunhung replied:
ok. I will remove it. thanks.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -345,6 +339,26 @@ static int mt8195_dsp_remove(struct snd_sof_dev *sdev)
+
+	mach = &sof_mt8195_mach;
+	if (!mach) {
+		dev_warn(sdev->dev, "warning: No matching ASoC machine driver found\n");
```
remove warning prefix

yaochunhung replied:
ok. I will remove it. thanks.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -231,6 +231,18 @@ static int adsp_shared_base_ioremap(struct platform_device *pdev, void *data)
+	adsp_bootup_addr = SRAM_PHYS_BASE_FROM_DSP_VIEW;
+	dev_info(sdev->dev, "[ADSP] HIFIxDSP boot from base : 0x%08X\n",
+		 adsp_bootup_addr);
+	hifixdsp_boot_sequence((void *)sdev, adsp_bootup_addr);
```
feels too good to be true that FW boot can never fail

ranj063 commented on sound/soc/sof/mediatek/adsp_pcm.c:
```diff
@@ -0,0 +1,237 @@
+		return -EINVAL;
+
+	if (!mem_pool->dram_pool)
+		return -EINVAL;
```
if (!mem_pool || !mem_pool->dram_pool)?

yaochunhung replied:
ok, I will modify it. thanks.

ranj063 commented on sound/soc/sof/mediatek/adsp_pcm.c:
```diff
@@ -0,0 +1,237 @@
+	if (!mem_pool)
+		return -EINVAL;
+
+	if (!mem_pool->dram_pool)
```
same here. combine the 2 if's?

yaochunhung replied:
ok, I will modify it. thanks.

ranj063 commented on sound/soc/sof/mediatek/adsp_pcm.c:
```diff
@@ -0,0 +1,237 @@
+		return -EINVAL;
+
+	if (!mem_pool->dram_pool)
+		return -EINVAL;
```
same here

yaochunhung replied:
ok, I will modify it. thanks.

ranj063 commented on sound/soc/sof/mediatek/adsp_pcm.c:
```diff
@@ -0,0 +1,237 @@
+
+	spcm = snd_sof_find_spcm_dai(scomp, rtd);
+	if (!spcm) {
+		dev_warn(sdev->dev, "%s warn: can't find PCM with DAI ID %d\n",
```
dev_err

yaochunhung replied:
ok, I will modify it. thanks.

ranj063 commented on sound/soc/sof/mediatek/adsp_pcm.c:
```diff
@@ -0,0 +1,237 @@
+					       &dmab->area,
+					       &dmab->addr,
+					       dma_bytes,
+					       PAGE_SIZE);
```
can you use 100chars per line please?

ranj063 commented on sound/soc/sof/mediatek/adsp_pcm.c:
```diff
@@ -0,0 +1,237 @@
+
+	spcm = snd_sof_find_spcm_dai(scomp, rtd);
+	if (!spcm) {
+		dev_warn(sdev->dev, "%s warn: can't find PCM with DAI ID %d\n",
```
dev_err

yaochunhung replied:
ok. I will modify it. thanks.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195-ipc.c:
```diff
@@ -0,0 +1,244 @@
+		if (reply.hdr.size != msg->reply_size) {
+			dev_err(sdev->dev, "error: reply expected %zu got %u bytes\n",
+				msg->reply_size, reply.hdr.size);
+			ret = -EINVAL;
```
don't you need to set reply_error and return at this point? why proceed and read the mailbox?

yaochunhung replied:
ipc_cmd in the link https://github.com/thesofproject/sof/blob/main/src/ipc/ipc3/handler.c#L1583. It also return message even reply.error not negative, and I check there are these codes in other platforms. I think it should be correct to proceed and read the mailbox. Thanks.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195-ipc.c:
```diff
@@ -0,0 +1,244 @@
+	name = kasprintf(GFP_KERNEL, "mbox%d", ch_info->id);
+	if (!name) {
+		dev_err(sdev->dev, "Failed to get mbox%d\n", ch_info->id);
+		return -EINVAL;
```
return -ENOMEM please and no error message

ranj063 replied:
and free name before returning from the function

yaochunhung replied:
ok. I will modify it. thanks.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195-ipc.c:
```diff
@@ -0,0 +1,244 @@
+
+static void mt8195_mbox_shutdown(struct mbox_chan *chan)
+{
+}
```
please add a comment if this is a TODO

yaochunhung replied:
ok. thanks

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -55,6 +55,40 @@ void __iomem *get_mbox_reg_base(u32 id)
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: failed to init adsp ipi device:%d\n",
+			ret);
+	}
```
remoce braces

ranj063 replied:
you can simply return mt8195_mbox_init(); here.

yaochunhung replied:
ok. I will modify it . thanks.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -331,9 +391,27 @@ static int mt8195_dsp_probe(struct snd_sof_dev *sdev)
+	ret = adsp_genpool_create(&priv->mem_pool, &share_dram);
+	if (ret) {
+		dev_err(sdev->dev, "adsp_genpool_create fail!\n");
+		ret = -ENOMEM;
```
return ret please

yaochunhung replied:
ok. I will modify it. thanks.

yaochunhung commented:
> @yaochunhung I've gone your patch series and added some comments. There is still work to be done. I would suggest you take it slower with simple patches first and then go for more delicate ones.
> 
> You can run git log --oneline sound/soc/sof/imx/ and see how we added our initial support for i.MX8 boards. It is not complete, we are now sending our clocks patches, then following by PM support, then the rest of the stuff.
> 
> If you want and consider it necessary we could also setup a meeting to discuss more.

dbaluta commented:
@yaochunhung my suggestion for this is to submit only the obvious correct patches and then add the rest later. The amount of patches right now will make the code hard the review. 
I would suggest on going with:
- adding sound/soc/sof/mtk/ -> stuff (MTK HW support only)
- sound-of-dev.c changes

Add PM support later.

yaochunhung commented:
> @yaochunhung my suggestion for this is to submit only the obvious correct patches and then add the rest later. The amount of patches right now will make the code hard the review.
> I would suggest on going with:
> 
> * adding sound/soc/sof/mtk/ -> stuff (MTK HW support only)
> * sound-of-dev.c changes
> 
> Add PM support later.
@dbaluta  Thanks for suggestion. Do you mean that I should separate my patches and only summit correct patches? Because PM runtime is also handled in probe function, should I remove it first? Thanks.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195-clk.c:
```diff
@@ -0,0 +1,186 @@
+	} else { /* Power down ADSP SRAM */
+		writel(readl(va_dspsysreg) | DSP_SRAM_POOL_PD_MASK,
+		       va_dspsysreg);
+	}
```
braces not needed

yaochunhung replied:
ok. I will remove them. thanks

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,359 @@
+		return ret;
+	}
+
+	dev_info(dev, "[ADSP][DMA] pbase=0x%llx, size=0x%llx\n",
```
is this really meant to be info? or can it be dbg?

yaochunhung replied:
ok. I think it can be dbg. thanks

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,359 @@
+		return -EINVAL;
+	}
+
+	dev_info(dev, "[ADSP] dram pbase=%pa, dramsize=0x%x\n",
```
same here, dbg level maybe?

yaochunhung replied:
ok. I will modify it. thanks

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,359 @@
+		return -EINVAL;
+	}
+
+	dev_info(dev, "[ADSP] sram pbase=%pa,0x%x\n", &adsp->pa_sram,
```
same here

yaochunhung replied:
ok. I will modify it. thanks

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,359 @@
+		if (IS_ERR(adsp->va_mboxreg[i]))
+			return PTR_ERR(adsp->va_mboxreg[i]);
+
+		dev_info(dev, "[ADSP]MBOX%d pa_mboxreg:%pa, va_mboxreg:%p\n",
```
could you please go through all th dev_info's in your series to make sure they need to be info and not dbg?

yaochunhung replied:
ok. I think they could be dbg level. thanks

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -0,0 +1,359 @@
+static int mt8195_get_bar_index(struct snd_sof_dev *sdev, u32 type)
+{
+	return type;
+}
```
is this a TODO? why return an argument as is?

yaochunhung replied:
It is one-to-one match between type and Bar index so it just return type. thanks

ranj063 replied:
can you please add a comment to say this?

yaochunhung replied:
ok. I will add comment for it. thanks

ranj063 commented on sound/soc/sof/mediatek/adsp_ipc.c:
```diff
@@ -0,0 +1,67 @@
+	ch_info->ipi_op_val = op;
+	ret = mbox_send_message(priv->ipi[idx].chan, NULL);
+	if (ret < 0)
+		dev_err(sdev->dev, "failed to send message via mbox: %d\n", ret);
```
if this can fail, should this function return the error?

yaochunhung replied:
ok . I will modify it. thanks.

ranj063 commented on sound/soc/sof/mediatek/mt8195/mt8195.c:
```diff
@@ -54,6 +55,32 @@ void __iomem *get_mbox_reg_base(u32 id)
+	sof_mailbox_write(sdev, sdev->host_box.offset, msg->msg_data,
+			  msg->msg_size);
+
+	adsp_ipi_send(sdev, ADSP_IPI_MBOX_REQ, ADSP_IPI_OP_REQ);
```
this can file, so ideally you should return the error to indicate if the msg TX failed

yaochunhung replied:
ok. I will add it. thanks.

yaochunhung commented:
Per discussion with @dbaluta , I will close this pull and keep it for reference. Then create new PR for basic skeleton  for MTK hardware support first. Thanks.

cujomalainey commented:
@yaochunhung please link the followup PRs here

yaochunhung commented:
> @yaochunhung please link the followup PRs here

@cujomalainey  ok. I will link them. thanks!

Basic hardware support for mt8195 : https://github.com/thesofproject/linux/pull/3160 (merged)
Add PCM stream callback for mt8195 : https://github.com/thesofproject/linux/pull/3190(Pending)
ASoC: SOF: pcm: add sof_keep_normal_fes to keep normal fes : https://github.com/thesofproject/linux/pull/3217
